### PR TITLE
fix(text): rich text overflow

### PR DESCRIPTION
**Pull request by Ovilia**
Support contentWidth for plainText so that apache/echarts#16034 could do the calculation.

### CODE REVIEW

File: src/graphic/helper/parseText.ts [hunk 1]
```diff
@@ -678,7 +689,6 @@ function wrapText(
                     // Append lastWord if have
                     if (currentWord) {
                         line += currentWord;
-                        accumWidth += currentWordWidth;
```
Comment (Ovilia) [inline src/graphic/helper/parseText.ts:681]: Each character's width has already added to `accumWidth` when looping through. Typical case:

**Before**
image
The third line's width is calculated wrong due to this line.

**After**
image

File: src/canvas/graphic.ts [hunk 2]
```diff
@@ -407,10 +407,8 @@ function bindCommonProps(
         }
     }
     if (forceSetAll || style.opacity !== prevStyle.opacity) {
-        if (!styleChanged) {
-            flushPathDrawn(ctx, scope);
-            styleChanged = true;
-        }
```
Comment (Ovilia) [inline src/canvas/graphic.ts:413]: Based on https://deepscan.io/dashboard/#view=project&tid=16023&pid=19250&bid=494172&subview=pull-request&prid=1016382

File: src/canvas/graphic.ts [hunk 3]
```diff
@@ -484,10 +482,10 @@ function bindPathAndTextCommonStyle(
         }
         ctx.globalAlpha = style.opacity == null ? 1 : style.opacity;
     }
-    if (el.hasStroke()) {
```
Comment (Ovilia) [inline src/canvas/graphic.ts:487]: Based on https://deepscan.io/dashboard/#view=project&tid=16023&pid=19250&bid=494172&subview=pull-request&prid=1016382

File: src/canvas/graphic.ts (hunk 2 above)
Comment (plainheart) [reply src/canvas/graphic.ts:413]: Caught an awesome project. Thanks!

File: src/canvas/graphic.ts (hunk 3 above)
Comment (pissang) [reply src/canvas/graphic.ts:487]: Seems I don't see it in the deepscan result. el can't be null here

File: src/graphic/Text.ts [hunk 4]
```diff
@@ -484,7 +484,12 @@ class ZRText extends Displayable<TextProps> implements GroupLike {
         const needDrawBg = needDrawBackground(style);
         const bgColorDrawn = !!(style.backgroundColor);
 
-        let outerHeight = contentBlock.outerHeight;
+        // outerWidth is the real text width containing padding.
+        const outerHeight = contentBlock.outerHeight;
+        const outerWidth = contentBlock.outerWidth;
+        const innerWidth = textPadding
```
Comment (pissang) [inline src/graphic/Text.ts:490]: Can we provide `contentWidth` like before and avoid calculating it with padding again?

File: src/canvas/graphic.ts (hunk 2 above)
Comment (pissang) [reply src/canvas/graphic.ts:413]: Impressive!

Comment (pissang) [reply src/canvas/graphic.ts:413]: Impressive!

File: src/graphic/Text.ts [hunk 5]
```diff
@@ -501,12 +503,10 @@ class ZRText extends Displayable<TextProps> implements GroupLike {
...
         if (needDrawBg || textPadding) {
             // Consider performance, do not call getTextWidth util necessary.
-            let outerWidth = contentBlock.width;
-            textPadding && (outerWidth += textPadding[1] + textPadding[3]);
-            const boxX = adjustTextX(baseX, outerWidth, textAlign);
+            const boxX = adjustTextX(baseX, contentWidth, textAlign);
             const boxY = adjustTextY(baseY, outerHeight, verticalAlign);
 
-            needDrawBg && this._renderBackground(style, style, boxX, boxY, outerWidth, outerHeight);
+            needDrawBg && this._renderBackground(style, style, boxX, boxY, contentWidth, outerHeight);
```
Comment (pissang) [inline src/graphic/Text.ts:509]: Is it a mistake to render the background with `contentWidth` instead of `outerWidth`, which will include the padding

**Comment by Jancheng-z** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我已经收到你的邮件，我将仔细阅读来信，祝君身体健康，合家美满！

**Review by pissang — APPROVED** [review]

**Pull request merged**